analysis/furtherModelTesting.ipynb

```diff
@@ -0,0 +1,1039 @@
+      "MAE for Decision Trees is: 1.0067465393990278\n"
+     ]
+    }
+   ],
+   "source": [
+    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=global_test_size, random_state=global_random_state)\n",
+    "\n",
+    "dt_model = DecisionTreeRegressor(random_state=global_random_state)\n",
+    "dt_model.fit(X_train, y_train)\n",
+    "\n",
+    "y_pred_dt = dt_model.predict(X_test)\n",
+    "\n",
+    "rmse_dt = np.sqrt(mean_squared_error(y_test, y_pred_dt))\n",
+    "mae_dt = mean_absolute_error(y_test, y_pred_dt)\n",
+    "\n",
+    "print(f\"RMSE for Decision Trees is: {rmse_dt}\")\n",
+    "print(f\"MAE for Decision Trees is: {mae_dt}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Decision Trees with Cross Validation"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 19,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Average RMSE of Decision Trees with Cross Validation is: 1.6194096052355795\n",
+      "Average MAE of Decision Trees with Cross Validation is: 0.7345414044039864\n"
+     ]
+    }
+   ],
+   "source": [
+    "kf = KFold(n_splits=5, shuffle=True, random_state=global_random_state)\n",
+    "dt_model = DecisionTreeRegressor(random_state=global_random_state)\n",
+    "\n",
+    "rmse_scores_dt = []\n",
+    "mae_scores_dt = []\n",
+    "\n",
+    "for train_index, val_index in kf.split(X):\n",
+    "    X_train, X_val = X.iloc[train_index], X.iloc[val_index]\n",
+    "    y_train, y_val = y.iloc[train_index], y.iloc[val_index]\n",
+    "    \n",
+    "    dt_model.fit(X_train, y_train)\n",
+    "    y_pred_dt = dt_model.predict(X_val)\n",
+    "    \n",
+    "    rmse_scores_dt.append(np.sqrt(mean_squared_error(y_val, y_pred_dt)))\n",
+    "    mae_scores_dt.append(mean_absolute_error(y_val, y_pred_dt))\n",
+    "\n",
+    "print(f\"Average RMSE of Decision Trees with Cross Validation is: {np.mean(rmse_scores_dt)}\")\n",
+    "print(f\"Average MAE of Decision Trees with Cross Validation is: {np.mean(mae_scores_dt)}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Random Forests"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 20,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Best hyperparameters for Random Forest: {'max_depth': 10, 'min_samples_leaf': 1, 'min_samples_split': 10, 'n_estimators': 100}\n",
+      "RMSE for Random Forests is: 1.6678055537819902\n",
+      "MAE for Random Forests is: 0.9348572302396694\n"
+     ]
+    }
+   ],
+   "source": [
+    "# HYPERPARAMETER TUNING FOR RANDOM FORESTS\n",
+    "\n",
+    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=global_test_size, random_state=global_random_state)\n",
+    "# Define the parameter grid for Random Forest\n",
+    "param_grid_rf = {\n",
```

Comment:
Tell me how many times this model will be trained with this size of param_grid

Reply:
How would this change if we also wanted to use Cross validation of say... 5?

Reply:
the model is trained once for each combination, so it's 3x4x3x3=108 times. cross validation of 5 means we would train 5 times per combination so 5x108= 540. Is this too much?